### PR TITLE
EventSetupRecord::doGet uses prefetching

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Changed EventSetupRecord::doGet to work with prefetching. This required the introduction of ESGetTokenGeneric and a new esConsumes method.
Added the EDConsumerBase::registerLateConsumes to be able to safely call esConsumes based on the contents of the EventSetup.

#### PR validation:

Code compiles and relevant framework unit tests pass.

### CODE REVIEW

File: FWCore/Modules/src/EventSetupRecordDataGetter.cc [hunk 1]
```diff
@@ -24,38 +24,50 @@
...
 #include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Utilities/interface/ESGetTokenGeneric.h"
 
 //
 // class decleration
 //
 namespace edm {
-  class EventSetupRecordDataGetter : public edm::stream::EDAnalyzer<> {
+  class EventSetupRecordDataGetter
+      : public edm::global::EDAnalyzer<edm::RunCache<std::nullptr_t>, edm::LuminosityBlockCache<std::nullptr_t>> {
```
Comment (Dr15Jones) [inline FWCore/Modules/src/EventSetupRecordDataGetter.cc:43]: As a bonus, I could make it a global module :).

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+1

**Comment by qliphy** [comment]
+1

**Pull request merged**